### PR TITLE
feat: property-based scheduling: add AKS property provider

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR adds the AKS property provider implementation.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests
- [x] Integration tests
- [x] E2E tests

### Special notes for your reviewer

Some tests are removed from this PR to control its size; will submit a new PR to add them back.

### CODE REVIEW

File: pkg/propertyprovider/aks/controllers/pod.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers
```
Comment (zhiying-lin) [inline pkg/propertyprovider/aks/controllers/pod.go:6]: do we have integration tests for these controllers in the future PR?

File: pkg/propertyprovider/aks/controllers/node.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"go.goms.io/fleet/pkg/propertyprovider/aks/trackers"
+)
+
+// NodeReconciler reconciles Node objects.
+type NodeReconciler struct {
+	NT     *trackers.NodeTracker
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/node.go:23]: Done.

File: pkg/propertyprovider/aks/controllers/node.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/node.go:6]: Done. Considering that the controllers here are quite simple, separate packages might not be necessary. If there are more object types to watch, we could change this, of course.

File: pkg/propertyprovider/aks/controllers/pod.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+// TO-DO (chenyu1): this is a relatively expensive watcher, due to how frequent pods can change
+// in a Kubernetes cluster; unfortunately at this moment there does not seem to be a better way
+// to observe the changes of requested resources in a cluster. The alternative, which is to use
+// Lists, adds too much overhead to the API server.
+
+// PodReconciler reconciles Pod objects.
+type PodReconciler struct {
+	PT     *trackers.PodTracker
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/pod.go:28]: Done.

File: pkg/propertyprovider/aks/controllers/pod.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			// the case, the removal (untracking) operation is a no-op.
+			//
+			// Note that this controller will not add any finalizer to pod objects, so as to
+			// avoid blocking normal Kuberneters operations under unexpected circumstances.
+			p.PT.Remove(req.NamespacedName.String())
+			return ctrl.Result{}, nil
+		}
+
+		// For other errors, retry the reconciliation.
+		klog.V(2).ErrorS(err, "Failed to get the pod object", "pod", podRef)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/pod.go:58]: Done.

File: pkg/propertyprovider/aks/controllers/node.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+type NodeReconciler struct {
+	NT     *trackers.NodeTracker
+	Client client.Client
+}
+
+// Reconcile reconciles a node object.
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	nodeRef := klog.KRef(req.Namespace, req.Name)
+	startTime := time.Now()
+	klog.V(2).InfoS("Reconciliation starts", "node", nodeRef)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/node.go:31]: Done. I rephrased the message a bit to be more clear.

File: pkg/propertyprovider/aks/controllers/node.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Reconciliation ends", "node", nodeRef, "latency", latency)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/node.go:34]: Done.

File: pkg/propertyprovider/aks/controllers/node.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+			// the case, the removal (untracking) operation is a no-op.
+			//
+			// Note that this controller will not add any finalizer to node objects, so as to
+			// avoid blocking normal Kuberneters operations under unexpected circumstances.
+			klog.V(2).InfoS("Node is not found; untrack it from the property provider", "node", nodeRef)
+			r.NT.Remove(req.Name)
+			return ctrl.Result{}, nil
+		}
+		// For other errors, retry the reconciliation.
+		klog.V(2).ErrorS(err, "Failed to get the node object", "node", nodeRef)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/node.go:53]: Done.

File: pkg/propertyprovider/aks/controllers/pod.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// Reconcile reconciles a pod object.
+func (p *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	podRef := klog.KRef(req.Namespace, req.Name)
+	startTime := time.Now()
+	klog.V(2).InfoS("Reconciliation starts", "pod", podRef)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Reconciliation ends", "pod", podRef, "latency", latency)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/pod.go:39]: Done.

File: pkg/propertyprovider/aks/provider.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+	PropertyCollectionSucceededReason                = "AllPropertiesCollectedSuccessfully"
+	PropertyCollectionFailedCostErrorReason          = "FailedToCollectCosts"
+	PropertyCollectionSucceededMessage               = "All properties have been collected successfully"
+	PropertyCollectionFailedCostErrorMessageTemplate = "An error has occurred when collecting cost properties: %v"
+)
+
+// PropertyProvider is the AKS property provider for Fleet.
+type PropertyProvider struct {
+	// The trackers.
+	pt *trackers.PodTracker
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:71]: Done.

File: pkg/propertyprovider/aks/provider.go [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+		// member agent wins the leader election. It should be noted that running the AKS property
+		// provider for multiple times will not incur any side effect other than some minor
+		// performance costs, as at this moment the AKS property provider observes data individually
+		// in a passive manner with no need for any centralized state.
+		LeaderElection: false,
+	})
+	p.mgr = mgr
+
+	if err != nil {
+		klog.Errorf("Failed to start AKS property provider: %v", err)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:118]: Done.

File: pkg/propertyprovider/aks/provider.go [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+		ObservationTime: metav1.Now(),
+	}
+
+	perCPUCost, perGBMemoryCost, err := p.nt.Costs()
+	if err != nil {
+		conds = append(conds, metav1.Condition{
+			Type:    PropertyCollectionSucceededConditionType,
+			Status:  metav1.ConditionFalse,
+			Reason:  "FailedToCollectCosts",
+			Message: fmt.Sprintf(PropertyCollectionFailedCostErrorMessageTemplate, err),
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:203]: Done.

File: pkg/propertyprovider/aks/suite_test.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"go.goms.io/fleet/pkg/propertyprovider"
+	"go.goms.io/fleet/pkg/propertyprovider/aks/trackers"
+)
+
+const (
+	eventuallyDuration = time.Second * 20
+	eventuallyInterval = time.Second * 5
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/suite_test.go:30]: Our tests use a eventuallyDuration between 10s to several minutes, and a eventuallyInterval between 10s of milliseconds to seconds. The values here are picked as the tests here need to do some calculations manually to verify the correctness of returned data, which can be a bit expensive; and this cross-ref. requires a stable state.

File: pkg/propertyprovider/aks/provider_test.go [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+func (d *dummyPricingProvider) OnDemandPrice(_ string) (float64, bool) {
+	return 1.0, true
+}
+
+func (d *dummyPricingProvider) LastUpdated() time.Time {
+	return currentTime
+}
+
+func TestCollect(t *testing.T) {
+	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider_test.go:67]: Oh, such tests are stripped out of this PR to control its size. They will be added back in a future PR.

File: pkg/propertyprovider/aks/provider.go [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+
+	if p.region == nil {
+		klog.V(2).Info("Auto-discover region as none has been specified")
+		// Note that an API reader is passed here for the purpose of auto-discovering region
+		// information from AKS nodes; at this time the cache from the controller manager
+		// has not been initialized yet and as a result cached client is not yet available.
+		//
+		// This incurs the slightly higher overhead, however, as auto-discovery runs only
+		// once, the performance impact is negligible.
+		if err := p.autoDiscoverRegionAndSetupTrackers(ctx, mgr.GetAPIReader()); err != nil {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:130]: Done.

File: pkg/propertyprovider/aks/provider.go [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+	return &PropertyProvider{
+		region: region,
+	}
+}
+
+// NewWithPricingProvider returns a new AKS property provider with the given
+// pricing provider.
+//
+// This is mostly used for testing purposes.
+func NewWithPricingProvider(pp trackers.PricingProvider) propertyprovider.PropertyProvider {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:327]: Oh, right. About this one, I know that we discussed this offline, but it just occurred to me that one of the reasons the method is added here as the pricing client might not necessarily be a Karpenter based one (which supports retail pricing only). I will edit the comment to be clearer about this

File: pkg/propertyprovider/aks/controllers/pod.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/propertyprovider/aks/controllers/pod.go:6]: Those controllers are quite straightforward so at this moment I feel that they might not need individual integration tests (as the provider ones already cover all of their branches). We could always add some, though, if necessary

**Review by ryanzhang-oss — COMMENTED** [review]
I think we need a few tests

**Review by ryanzhang-oss — APPROVED** [review]

File: pkg/propertyprovider/aks/provider.go [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+	//
+	// An alternative is to perform a list for once during the startup, which might be
+	// too expensive for a large cluster.
+	mgr.GetCache().WaitForCacheSync(ctx)
+
+	return nil
+}
+
+// Collect collects the properties of an AKS cluster.
+func (p *PropertyProvider) Collect(_ context.Context) propertyprovider.PropertyCollectionResponse {
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/propertyprovider/aks/provider.go:189]: Hi Ryan! This is not used as at this moment there's no long running ops for this Collect() call (all the data are collected asynchronously). I could write a timeout check here, but the upper layer is already checking for that, so it feels a bit redundant.

**Comment by michaelawyu** [comment]
Will merge this PR to unblock the progress; if there's any concern, please let me know. As mentioned, some tests will be submitted in a separate PR.

**Pull request merged**